### PR TITLE
GCS support

### PR DESCRIPTION
**Pull request by tristanvanthielen**
A first attempt at adding GCS support to Terraboard. Allows authentication either using GOOGLE_APPLICATION_CREDENTIALS or by mounting a service account key and specifying the path as an argument.

I tried to match S3 design as much as possible.

Requirements:
- Versioning should be enabled on the bucket
- Service account must have read access on the specified bucket

Tested:
- Was able to import versioned state files from GCS and view diffs between them as normal
- Connecting both using default credentials and service accounts.

Should still be tested:
- Proper display of locked states. Did not have a good setup for testing this.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -99,7 +99,7 @@ func refreshDB(syncInterval uint16, d *db.Database, sp state.Provider) {
 						"path":       st,
 						"version_id": v.ID,
 						"error":      err,
-					}).Error("Failed to fetch state from S3")
+					}).Error("Failed to fetch state")
```
Comment (raphink) [inline main.go:102]:
```suggestion
					}).Error("Failed to fetch state from bucket")
```

**Comment by raphink** [comment]
Looks pretty good to me, good job!

Let me know when you've been able to test the whole thing, incl. locks, and I'll be happy to merge it!

**Comment by naiduarvind** [comment]
Any update on it? Keen to use Terraboard. Let me know if I can provide any kind of assistance to get this merged and ready for PRR.

**Comment by raphink** [comment]
@naiduarvind can you confirm that this patch works well with locks? That would help!

**Comment by naiduarvind** [comment]
@raphink testing it and will update here once done.

File: README.md [hunk 2]
```diff
@@ -119,6 +119,8 @@ The precedence of configurations is as described below.
 |`--tfe-address` | `TFE_ADDRESS` | `tfe.tfe-address` | Terraform Enterprise address for states access | - |
 |`--tfe-token` | `TFE_TOKEN` | `tfe.tfe-token` | Terraform Enterprise token for states access | - |
 |`--tfe-organization` | `TFE_ORGANIZATION` | `tfe.tfe-organization` | Terraform Enterprise organization for states access | - |
+|`--gcs-bucket` | `N/A` | `gcp.gcs-buckets` | Google Cloud Storage buckets to access | - |
```
Comment (naiduarvind) [inline README.md:122]: @tristanvanthielen - are we missing the variable for the GCS bucket?

**Comment by naiduarvind** [comment]
@raphink - could you point me to some reference on building and running terraboard with @tristanvanthielen's changes?

I ran `docker-compose build && docker-compose up -d` but hit an error locally:

Screenshot 2020-05-20 at 12 10 43 AM

File: README.md (hunk 2 above)
Comment (tristanvanthielen) [reply README.md:122]: Hi, so I did not include a variable as the gcs-bucket argument can be specified multiple times to pull states from multiple buckets. I did not really find a clean way to solve this using env variables, but if you have an idea, feel free to suggest

**Comment by naiduarvind** [comment]
@tristanvanthielen a little bit of guide on how you got terraboard running locally. I am running a Mac and hitting issues with clang. Do you have similar issues as well?

Screenshot 2020-05-20 at 12 24 30 AM

File: README.md (hunk 2 above)
Comment (raphink) [reply README.md:122]: `go-flags` allows to pass lists in environment variables, using the `env-delim` argument for the delimiter, for example:

```go
  GCSBuckets []string `long:"gcs-bucket" env:"GCS_BUCKET" env-delim:","`
```

**Comment by raphink** [comment]
@naiduarvind `docker-compose build` actually works fine for me.

In your last screenshot, the error indicates you're missing the `crt0` dev libs to build. I don't have a build env on a Mac at the moment to check.

https://stackoverflow.com/questions/3801011/ld-library-not-found-for-lcrt0-o-on-osx-10-6-with-gcc-clang-static-flag/3801032 suggests static compilation might not be possible on MacOS actually, you could could just try dynamic building with `go build` instead.

**Comment by raphink** [comment]
What's the status of this PR? Have you been able to test it enough?

**Pull request merged**